pkg/controller/workspace/component/plugins_utils.go

```diff
@@ -0,0 +1,266 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package component
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	workspaceApi "github.com/che-incubator/che-workspace-operator/pkg/apis/workspace/v1alpha1"
+	k8sModelUtils "github.com/che-incubator/che-workspace-operator/pkg/controller/modelutils/k8s"
+	"github.com/che-incubator/che-workspace-operator/pkg/controller/workspace/config"
+	"github.com/che-incubator/che-workspace-operator/pkg/controller/workspace/model"
+	"github.com/che-incubator/che-workspace-operator/pkg/controller/workspace/server"
+	brokerModel "github.com/eclipse/che-plugin-broker/model"
+	"github.com/eclipse/che-plugin-broker/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func convertToComponentInstanceStatus(plugin brokerModel.ChePlugin, wkspCtx model.WorkspaceContext) (*model.ComponentInstanceStatus, error) {
+	pod, err := createPodFromPlugin(plugin, wkspCtx)
+	if err != nil {
+		return nil, err
+	}
+	services := createServicesFromPlugin(plugin, wkspCtx)
+	var externalObjects []runtime.Object
+	for _, service := range services {
+		externalObjects = append(externalObjects, service)
+	}
+	endpoints := createEndpointsFromPlugin(plugin)
+	commands := createCommandsFromPlugin(plugin, wkspCtx)
+	containerDescriptions := createDescriptionsFromPlugin(plugin)
+
+	component := &model.ComponentInstanceStatus{
+		WorkspacePodAdditions:      pod,
+		ExternalObjects:            externalObjects,
+		Endpoints:                  endpoints,
+		ContributedRuntimeCommands: commands,
+		Containers:                 containerDescriptions,
+	}
+
+	return component, nil
+}
+
+func createPodFromPlugin(plugin brokerModel.ChePlugin, wkspCtx model.WorkspaceContext) (*corev1.PodTemplateSpec, error) {
```

Comment:
This file brings a lot of package visible methods in `controller/workspace/component` package.
Maybe it makes sense to move `plugins` and `plugins_utils` in its own package like `controller/workspace/component/plugin` ?

Reply:
It's tricky to do without a quite large refactor, since the plugins provisioning code depends on functions in `component`; without moving all utility functions in `component` elsewhere, we get an import cycle `component -> component/plugins -> component`